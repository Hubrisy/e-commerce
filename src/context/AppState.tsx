import type { Dispatch, PropsWithChildren, SetStateAction } from 'react';
import { createContext, useContext } from 'react';

import { useStorageValue } from '@/hooks/use-storage-value';
import { StorageKeys } from '@/storage/localstorage';
import type { UserData } from '@/types';

interface AppStateTypes {
  favorites: Array<number>;
  user: UserData;
}

interface AppStateContextTypes extends AppStateTypes {
  setUser: Dispatch<SetStateAction<AppStateTypes['user']>>;
  setFavorites: Dispatch<SetStateAction<AppStateTypes['favorites']>>;
}

const defaultUser: UserData = {
  name: '',
  secondName: '',
  phone: '',
  email: '',
  city: '',
  address: '',
};

const AppStateContext = createContext<AppStateContextTypes | undefined>(
  undefined,
);

export const AppStateContextProvider: React.FC<PropsWithChildren> = ({
  children,
}) => {
  const [user, setUser] = useStorageValue<AppStateTypes['user']>(
    StorageKeys.user,
    defaultUser,
  );

  const [favorites, setFavorites] = useStorageValue<AppStateTypes['favorites']>(
    StorageKeys.favorites,
    [],
  );

  return (
    <AppStateContext.Provider
      value={{
        user,
        setUser,
        favorites,
        setFavorites,
      }}
    >
      {children}
    </AppStateContext.Provider>
  );
};

export const useAppStateContext = () => {
  const value = useContext(AppStateContext);

  if (!value) {
    throw new Error(
      'useAppStateContext should be called inside AppStateContextProvider',
    );
  }

  return value;
};
