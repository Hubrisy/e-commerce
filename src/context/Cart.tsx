import {
  createContext,
  type Dispatch,
  type PropsWithChildren,
  type SetStateAction,
  useContext,
  useState,
} from 'react';

import { useStorageValue } from '@/hooks/use-storage-value';
import { StorageKeys } from '@/storage/localstorage';
import type { CartItem, ShippingProduct } from '@/types';

interface CartTypes {
  cart: Array<CartItem>;
  selectedShipping?: ShippingProduct;
}

interface CartContextTypes extends CartTypes {
  setCart: Dispatch<SetStateAction<CartTypes['cart']>>;
  addToCart: (product: CartItem) => void;
  deleteFromCart: (id: number) => void;
  setSelectedShipping: Dispatch<SetStateAction<CartTypes['selectedShipping']>>;
}

const defaultCartState: CartTypes = {
  cart: [],
  selectedShipping: undefined,
};

const CartContext = createContext<CartContextTypes | undefined>(undefined);

export const CartContextProvider: React.FC<PropsWithChildren> = ({
  children,
}) => {
  const [cart, setCart] = useStorageValue<CartTypes['cart']>(
    StorageKeys.cart,
    defaultCartState.cart,
  );

  const [selectedShipping, setSelectedShipping] =
    useState<CartTypes['selectedShipping']>(undefined);

  const addToCart = (product: CartItem) => {
    const productAlreadyInCart = cart.find(item => item.id === product.id);

    if (productAlreadyInCart) {
      return;
    }

    const updateCart = [...cart, product];

    if (updateCart.length) {
      setCart(updateCart);
    }
  };

  const deleteFromCart = (id: number) => {
    const updateCart = cart.filter(item => item.id !== id);
    setCart(updateCart);
  };

  return (
    <CartContext.Provider
      value={{
        cart,
        setCart,
        addToCart,
        deleteFromCart,
        selectedShipping,
        setSelectedShipping,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export const useCartContext = () => {
  const value = useContext(CartContext);

  if (!value) {
    throw Error('useCartContext should be called inside CartContextProvider');
  }

  return value;
};
