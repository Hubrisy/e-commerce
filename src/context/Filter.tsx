import {
  createContext,
  type Dispatch,
  type PropsWithChildren,
  type SetStateAction,
  useContext,
  useState,
} from 'react';

import type { ProductCategory, ProductFeature } from '@/types';

interface FilterState {
  category?: ProductCategory;
  feature?: ProductFeature;
}
interface FilterContextType extends FilterState {
  setCategory: Dispatch<SetStateAction<FilterState['category']>>;
  setFeature: Dispatch<SetStateAction<FilterState['feature']>>;
}

const defaultFilterState: FilterState = {
  category: undefined,
  feature: undefined,
};

const FilterContext = createContext<FilterContextType | undefined>(undefined);

export const FilterContextProvider: React.FC<PropsWithChildren> = ({
  children,
}) => {
  const [category, setCategory] = useState<FilterState['category']>(
    defaultFilterState.category,
  );
  const [feature, setFeature] = useState<FilterState['feature']>(
    defaultFilterState.feature,
  );

  return (
    <FilterContext.Provider
      value={{ category, setCategory, feature, setFeature }}
    >
      {children}
    </FilterContext.Provider>
  );
};

export const useFilterContext = () => {
  const value = useContext(FilterContext);

  if (!value) {
    throw Error(
      'useFilterContext should be called inside FilterContextProvider',
    );
  }

  return value;
};
