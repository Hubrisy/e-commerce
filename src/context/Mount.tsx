import {
  createContext,
  type PropsWithChildren,
  useContext,
  useEffect,
  useState,
} from 'react';

interface MountDataType {
  isMounted: boolean;
}

const defaultState: MountDataType = {
  isMounted: false,
};

const MountContext = createContext<MountDataType | undefined>(undefined);

export const MountContextProvider: React.FC<PropsWithChildren> = ({
  children,
}) => {
  const [isMounted, setIsMounted] = useState<MountDataType['isMounted']>(
    defaultState.isMounted,
  );

  useEffect(() => {
    setIsMounted(true);
  }, []);

  return (
    <MountContext.Provider value={{ isMounted }}>
      {children}
    </MountContext.Provider>
  );
};

export const useMountContext = () => {
  const value = useContext(MountContext);

  if (!value) {
    throw new Error(
      'useMountContext should be called inside MountContextProvider',
    );
  }

  return value;
};
