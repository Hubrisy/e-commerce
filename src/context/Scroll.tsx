import {
  createContext,
  type PropsWithChildren,
  useCallback,
  useContext,
  useRef,
} from 'react';

interface ScrollTypes {
  shopRef?: React.RefObject<HTMLDivElement | null>;
  scrollToShop: () => void;
}

const ScrollContext = createContext<ScrollTypes | null>(null);

export const ScrollContextProvider: React.FC<PropsWithChildren> = ({
  children,
}) => {
  const shopRef = useRef<HTMLDivElement | null>(null);

  const scrollToShop = useCallback(() => {
    shopRef.current?.scrollIntoView({ behavior: 'smooth' });
  }, []);

  return (
    <ScrollContext.Provider value={{ shopRef, scrollToShop }}>
      {children}
    </ScrollContext.Provider>
  );
};

export const useScrollContext = () => {
  const context = useContext(ScrollContext);

  if (!context) {
    throw new Error(
      'useScrollContext must be used within a ScrollContextProvider',
    );
  }

  return context;
};
