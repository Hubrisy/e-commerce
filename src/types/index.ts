export type Currency = 'USD';

export const currencySymbols: Record<Currency, string> = {
  USD: '$',
};

export type ProductCategory =
  | 'phones'
  | 'smartwatches'
  | 'cameras'
  | 'headphones'
  | 'computers'
  | 'gaming'
  | 'tablet';

export type ProductFeature = 'newarrival' | 'bestseller' | 'featured';

export interface Product {
  name: string;
  img: string;
  price: number;
  currency: Currency;
  category: ProductCategory;
  feature?: ProductFeature;
  id: number;
  isOnsale?: boolean;
  isOnMainScreen?: boolean;
}

export enum SectionIds {
  products = 'products',
}

export interface CartItem extends Product {
  quantity: number;
}
