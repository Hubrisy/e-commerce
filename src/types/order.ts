import type { CartItem } from '.';

export interface OrderModel {
  products: Array<CartItem>;
}

export interface OrderModelWithId extends OrderModel {
  id: number;
}
