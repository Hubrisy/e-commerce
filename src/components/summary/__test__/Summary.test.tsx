import { render } from '@testing-library/react';

import { Summary } from '..';

import { defaultUser, useAppStateContext } from '@/context/AppState';
import { useCartContext } from '@/context/Cart';
import { useMountContext } from '@/context/Mount';
import { type CartItem, currencySymbols, type ShippingProduct } from '@/types';

jest.mock('@/context/Mount');
jest.mock('@/context/Cart');

jest.mock('@/context/AppState', () => ({
  ...jest.requireActual('@/context/AppState'),
  useAppStateContext: jest.fn(),
}));

jest.mock('@/components/Loader', () => ({
  Loader: () => <div data-testid="loader">Loading...</div>,
}));

const mockedUseMountContext = useMountContext as jest.MockedFunction<
  typeof useMountContext
>;

const mockedUseCartContext = useCartContext as jest.MockedFunction<
  typeof useCartContext
>;

const mockedUseAppStateContext = useAppStateContext as jest.MockedFunction<
  typeof useAppStateContext
>;

describe('Summary Component', () => {
  const mockedUser: Partial<typeof defaultUser> = {
    address: '123 Test St, Test City, TC 12345',
  };

  const mockedProduct: Partial<CartItem> = {
    id: 1,
    name: 'Test Product',
    img: '/test-product.jpg',
    price: 99.99,
    quantity: 2,
  };

  const mockedCart: Array<Partial<CartItem>> = [mockedProduct];

  const mockedSelectedShipping: Partial<ShippingProduct> = {
    id: 1,
    type: 'Standard Shipping',
    price: 5.99,
    description: 'Delivery in 5-7 business days',
    date: '2023-10-01',
  };

  beforeEach(() => {
    mockedUseMountContext.mockReturnValue({
      isMounted: true,
    } as unknown as ReturnType<typeof useMountContext>);

    mockedUseCartContext.mockReturnValue({
      cart: mockedCart,
      selectedShipping: mockedSelectedShipping,
    } as unknown as ReturnType<typeof useCartContext>);

    mockedUseAppStateContext.mockReturnValue({
      user: {
        ...defaultUser,
        ...mockedUser,
      },
    } as unknown as ReturnType<typeof useAppStateContext>);
  });

  test('renders Loader when not mounted', () => {
    mockedUseMountContext.mockReturnValue({
      isMounted: false,
    } as unknown as ReturnType<typeof useMountContext>);

    const { getByTestId } = render(<Summary />);
    expect(getByTestId('loader')).toBeInTheDocument();
  });

  test('Loader matches snapshot when not mounted', () => {
    mockedUseMountContext.mockReturnValue({
      isMounted: false,
    } as unknown as ReturnType<typeof useMountContext>);

    const { asFragment } = render(<Summary />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Summary correctly', () => {
    const { getByText, getByAltText } = render(<Summary />);

    const productImg = getByAltText(mockedProduct.name as string);
    expect(productImg).toBeInTheDocument();
    expect(productImg).toHaveAttribute('src', mockedProduct.img as string);
    expect(getByText(mockedProduct.name as string)).toBeInTheDocument();

    const productPriceText = `${currencySymbols.USD}${mockedProduct.price}`;
    expect(getByText(productPriceText)).toBeInTheDocument();

    const address = getByText(mockedUser.address as string);
    expect(address).toBeInTheDocument();

    const shippingMethod = getByText(mockedSelectedShipping.type as string);
    expect(shippingMethod).toBeInTheDocument();

    const subtotal = mockedCart.reduce(
      (acc, { price = 0, quantity = 1 }) => acc + price * quantity,
      0,
    );
    const subtotalText = `${currencySymbols.USD}${subtotal}`;
    expect(getByText(subtotalText)).toBeInTheDocument();

    const shippingPrice = mockedSelectedShipping.price || 0;
    const shippingPriceText = `${currencySymbols.USD}${shippingPrice}`;
    expect(getByText(shippingPriceText)).toBeInTheDocument();

    const total = subtotal + shippingPrice;
    const totalText = `${currencySymbols.USD}${total}`;
    expect(getByText(totalText)).toBeInTheDocument();
  });

  test('Summary matches snapshot', () => {
    const { asFragment } = render(<Summary />);
    expect(asFragment()).toMatchSnapshot();
  });
});
