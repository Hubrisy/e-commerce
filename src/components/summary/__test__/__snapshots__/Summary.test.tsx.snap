// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary Component Loader matches snapshot when not mounted 1`] = `
<DocumentFragment>
  <div
    class="mt-[128px]"
  >
    <div
      data-testid="loader"
    >
      Loading...
    </div>
  </div>
</DocumentFragment>
`;

exports[`Summary Component Summary matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="my-12 border-[#EBEBEB] rounded-xl border-[1px] lg:max-w-full xl:min-w-[550px]"
  >
    <div
      class="max-w-[90%] m-auto my-8"
    >
      <div
        class="text-[20px] font-medium"
      >
        Summary
      </div>
      <div
        class="my-6 text-[16px] font-medium"
      >
        <div
          class="rounded-xl bg-[#F6F6F6] mb-4 flex p-2 justify-between items-center text-center"
        >
          <img
            alt="Test Product"
            height="40"
            src="/test-product.jpg"
            width="40"
          />
          <div>
            Test Product
          </div>
          <div>
            $99.99
          </div>
          <div>
            Qnt: 2
          </div>
        </div>
      </div>
      <div>
        <div>
          <div
            class="font-medium text-[14px]"
          >
            Address
          </div>
          <div
            class="font-normal text-[16px] mt-2"
          >
            123 Test St, Test City, TC 12345
          </div>
        </div>
        <div
          class="mt-4"
        >
          <div
            class="font-medium text-[14px]"
          >
            Shippment method
          </div>
          <div
            class="font-normal text-[16px] mt-2"
          >
            Standard Shipping
          </div>
        </div>
      </div>
      <div
        class="mt-6 text-[16px]"
      >
        <div
          class="flex justify-between font-medium"
        >
          <div>
            Subtotal
          </div>
          <div>
            $199.98
          </div>
        </div>
        <div
          class="flex justify-between font-normal my-4"
        >
          <div>
            Estimated shipping & Handling
          </div>
          <div>
            $5.99
          </div>
        </div>
        <div
          class="flex justify-between font-medium"
        >
          <div>
            Total
          </div>
          <div>
            $205.97
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
